### ISSUE TITLE
Possible API Breaking Changes

### ISSUE THREAD

**Issue by nibanks** [microsoft/msquic#1709]
This Issue is here to track "nice to have" API changes, but require a breaking change to the API (i.e. would require a major version number increase). Since breaking changes are avoided as much as possible we will wait until we have "enough" things are worth making the breaking change for in one big batch.

1. Make `StreamStart` to be async by default. Perhaps even remove blocking version all together.
2. Make `SetParam`/`GetParam` values non-overlapping to catch/prevent code bugs (#1705). Perhaps even combine level and parameter into one field?

**Comment by geoffkizer** [comment]
It would be nice if ListenerStop/ListenerClose could work asynchronously, like their counterparts for connection and stream, and generate a SHUTDOWN_COMPLETE event when complete.

For .NET we need to call these from a finalizer, and in general we want to avoid blocking in finalizer calls (all finalizer calls happen on a single finalizer thread). The amount of blocking that occurs here is probably very small in practice, but it would still be good to avoid if possible.

It would also be nice to have this work the same way stream and connection work, because it simplifies our handle management and cleanup logic so that it's consistent across all three. Currently we have to do something a bit different for Listener because the close blocks.

**Comment by DemiMarie** [comment]
> This Issue is here to track "nice to have" API changes, but require a breaking change to the API (i.e. would require a major version number increase). Since breaking changes are avoided as much as possible we will wait until we have "enough" things are worth making the breaking change for in one big batch.

Could MsQuic support both versions of the API via different API function pointer tables?

**Comment by nibanks** [comment]
@DemiMarie yes, that's exactly what we'd do, but there is overhead to supporting multiple in parallel, plus it requires a code change from all existing clients to upgrade to the new version. So we are delaying this until there is significant reason.

**Comment by TYoungSL** [comment]
`QUIC_DATAGRAM_SEND_STATE` has `QUIC_DATAGRAM_SEND_SENT` as 0.
Would prefer bumping them all up 1 and adding `QUIC_DATAGRAM_SEND_UNKNOWN` as member 0.
Helps with using the enum as a type when tracking state and generated bindings too.

**Comment by TYoungSL** [comment]
I'd like to request unifying `QUIC_STATUS_` returns and callback returns to be platform and crypto-lib agnostic.

e.g. `_PENDING` could be -2 on posix platforms, 0x000703E5 on windows for schannel (and openssl?) or for windows kernel 0x103

**Comment by TYoungSL** [comment]
I'd like to propose a counter https://github.com/microsoft/msquic/issues/404 and revert https://github.com/microsoft/msquic/pull/1176.

Binding creation for libraries has had to deal with `sockaddr` as an opaque struct for many libraries for a long time.

The same arguments for having pass through platform specific status codes over uniform status codes applies here too.
There is more surface area that is not necessary for MsQuic to cover. It creates edge cases, such as Mac/BSD (30) not having the same `AF_INET6` for IPv6 as WinSock (23) or Linux/Posix (10).

Binding libraries can trivially create a small platform specific C library to perform a conversion from a string or other uniform type of the binding author's choice to a platform specific `sockaddr` or write the implementation in the language for the binding.

Examples of this are bindings to libuv which provides string conversion helpers for `sockaddr` opaquely.

**Comment by thhous-msft** [comment]
From our end above, reverting #1176 is trivial. We can just change QUIC_ADDRESS_FAMILY_IPV6 to AF_IPV6 directly. That would be the only like required from our end. And then from the .NET side they would just have to change their definition as well. So its a very easy to fix change.
